File: legacy-python/bot/handlers/schedule.py
"""Ввод расписания в личке: /schedule, /import, /wizard, /myschedule, /clear, /busy."""

from __future__ import annotations

import html
import logging
import re

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.constants import ParseMode
from telegram.ext import (
    CallbackQueryHandler,
    CommandHandler,
    ContextTypes,
    ConversationHandler,
    MessageHandler,
    filters,
)

from qairu.core.intervals import fmt_minutes, merge
from qairu.core.parser import ParseResult, _find_ranges, find_kind, is_all_day, parse_any
from qairu.db import repo
from qairu.i18n import t, weekday_name
from ..utils import chat_tz, parse_date_token, resolve_lang, sync_user, today_in

log = logging.getLogger(__name__)

# состояния диалогов
IMPORT_WAIT, IMPORT_CONFIRM = range(2)
WIZ_PICK_DAY, WIZ_WAIT_TEXT = range(2, 4)
BUSY_WAIT = 4

PRIVATE = filters.ChatType.PRIVATE


# --------------------------------------------------------------------------
# Импорт расписания текстом
# --------------------------------------------------------------------------

async def cmd_schedule(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    async with repo.session() as session:
        await sync_user(session, update.effective_user)
        lang = await resolve_lang(update, session)
        await session.commit()
    await update.effective_message.reply_text(t(lang, "schedule_intro"), parse_mode=ParseMode.HTML)
    return IMPORT_WAIT


async def on_schedule_text(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    message = update.effective_message
    text = message.text or ""

    async with repo.session() as session:
        await sync_user(session, update.effective_user)
        lang = await resolve_lang(update, session)
        await session.commit()

    result = parse_any(text)
    if not result.ok:
        await message.reply_text(t(lang, "parse_failed"), parse_mode=ParseMode.HTML)
        return IMPORT_WAIT

    context.user_data["pending_schedule"] = {
        "slots": [(s.weekday, s.start_min, s.end_min, s.label, s.parity, s.kind)
                  for s in result.slots],
        "free_days": result.free_days,
    }

    preview = _render_parse_preview(result, lang)
    body = t(lang, "parse_preview", schedule=preview)
    if result.errors:
        body += t(lang, "parse_errors", lines=html.escape("\n".join(result.errors[:5])))

    keyboard = InlineKeyboardMarkup(
        [[
            InlineKeyboardButton(t(lang, "btn_confirm"), callback_data="sch:ok"),
            InlineKeyboardButton(t(lang, "btn_retry"), callback_data="sch:retry"),
        ]]
    )
    await message.reply_text(body, parse_mode=ParseMode.HTML, reply_markup=keyboard)
    return IMPORT_CONFIRM


def _slot_suffix(lang: str, label: str, parity: int | None, kind: str) -> str:
    parts: list[str] = []
    if label:
        parts.append(html.escape(label))
    if parity is not None:
        parts.append(t(lang, "parity_odd" if parity == 0 else "parity_even"))
    if kind and kind != "class":
        parts.append(t(lang, f"kind_{kind}"))
    return f" <i>({', '.join(parts)})</i>" if parts else ""


def _render_parse_preview(result: ParseResult, lang: str) -> str:
    by_day: dict[int, list[tuple[int, int, str, int | None, str]]] = {}
    for slot in result.slots:
        by_day.setdefault(slot.weekday, []).append(
            (slot.start_min, slot.end_min, slot.label, slot.parity, slot.kind)
        )

    lines: list[str] = []
    for weekday in range(7):
        if weekday in by_day:
            items = sorted(by_day[weekday], key=lambda row: (row[0], row[1]))
            rendered = ", ".join(
                f"{fmt_minutes(start)}–{fmt_minutes(end)}"
                + _slot_suffix(lang, label, parity, kind)
                for start, end, label, parity, kind in items
            )
            lines.append(f"<b>{weekday_name(lang, weekday)}</b>: {rendered}")
        elif weekday in result.free_days:
            lines.append(f"<b>{weekday_name(lang, weekday)}</b>: {t(lang, 'day_free')}")
    return "\n".join(lines) or "—"


async def on_schedule_confirm(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    query = update.callback_query
    await query.answer()
    action = query.data.split(":", 1)[1]

    async with repo.session() as session:
        lang = await resolve_lang(update, session)

    if action == "retry":
        context.user_data.pop("pending_schedule", None)
        await query.edit_message_text(t(lang, "schedule_intro"), parse_mode=ParseMode.HTML)
        return IMPORT_WAIT

    pending = context.user_data.pop("pending_schedule", None)
    if not pending:
        await query.edit_message_text(t(lang, "error_generic"))
        return ConversationHandler.END

    touched_days = sorted({slot[0] for slot in pending["slots"]} | set(pending["free_days"]))
    async with repo.session() as session:
        await repo.replace_weekly_slots(
            session, update.effective_user.id, touched_days, pending["slots"], source="import"
        )
        await session.commit()

    await query.edit_message_text(t(lang, "schedule_saved"))
    return ConversationHandler.END


# --------------------------------------------------------------------------
# Мастер по дням
# --------------------------------------------------------------------------

def _day_keyboard(lang: str) -> InlineKeyboardMarkup:
    rows, row = [], []
    for weekday in range(7):
        row.append(InlineKeyboardButton(weekday_name(lang, weekday)[:3], callback_data=f"wiz:day:{weekday}"))
        if len(row) == 4:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    rows.append([InlineKeyboardButton(t(lang, "btn_done"), callback_data="wiz:done")])
    return InlineKeyboardMarkup(rows)


async def cmd_wizard(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    async with repo.session() as session:
        await sync_user(session, update.effective_user)
        lang = await resolve_lang(update, session)
        await session.commit()
    await update.effective_message.reply_text(
        t(lang, "wizard_pick_day"), reply_markup=_day_keyboard(lang)
    )
    return WIZ_PICK_DAY


async def on_wizard_day(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    query = update.callback_query
    await query.answer()
    async with repo.session() as session:
        lang = await resolve_lang(update, session)

    if query.data == "wiz:done":
        async with repo.session() as session:
            await repo.mark_filled(session, update.effective_user.id, True)
            await session.commit()
        await query.edit_message_text(t(lang, "wizard_done"))
        return ConversationHandler.END

    weekday = int(query.data.rsplit(":", 1)[1])
    context.user_data["wiz_day"] = weekday
    await query.edit_message_text(
        t(lang, "wizard_ask_day", day=weekday_name(lang, weekday)), parse_mode=ParseMode.HTML
    )
    return WIZ_WAIT_TEXT


async def on_wizard_text(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    message = update.effective_message
    weekday = context.user_data.get("wiz_day")
    async with repo.session() as session:
        lang = await resolve_lang(update, session)

    if weekday is None:
        await message.reply_text(t(lang, "error_generic"))
        return ConversationHandler.END

    ranges = _find_ranges(message.text or "")
    slots = [(weekday, start, end, "") for start, end, _, _ in ranges]

    async with repo.session() as session:
        await repo.replace_weekly_slots(session, update.effective_user.id, [weekday], slots, source="wizard")
        await session.commit()

    rendered = ", ".join(f"{fmt_minutes(s)}–{fmt_minutes(e)}" for _, s, e, _ in slots) or t(lang, "day_free")
    await message.reply_text(
        t(lang, "wizard_day_saved", day=weekday_name(lang, weekday), slots=rendered)
    )
    await message.reply_text(t(lang, "wizard_pick_day"), reply_markup=_day_keyboard(lang))
    return WIZ_PICK_DAY


# --------------------------------------------------------------------------
# Просмотр, очистка, разовая занятость
# --------------------------------------------------------------------------

async def cmd_my_schedule(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    async with repo.session() as session:
        await sync_user(session, update.effective_user)
        lang = await resolve_lang(update, session)
        slots = await repo.get_slots(session, update.effective_user.id)
        await session.commit()

    if not slots:
        await update.effective_message.reply_text(t(lang, "schedule_empty"))
        return

    weekly: dict[int, list[tuple[int, int, str, int | None, str]]] = {}
    dated: list[str] = []
    for slot in slots:
        if slot.specific_date is not None:
            dated.append(
                f"📌 <b>{slot.specific_date.strftime('%d.%m.%Y')}</b>: "
                f"{fmt_minutes(slot.start_min)}–{fmt_minutes(slot.end_min)}"
                + _slot_suffix(lang, slot.label, None, slot.kind)
            )
        elif slot.date_from is not None and slot.date_to is not None:
            dated.append(
                f"📌 <b>{slot.date_from.strftime('%d.%m')}–{slot.date_to.strftime('%d.%m.%Y')}</b>: "
                f"{fmt_minutes(slot.start_min)}–{fmt_minutes(slot.end_min)}"
                + _slot_suffix(lang, slot.label, None, slot.kind)
            )
        elif slot.weekday is not None:
            weekly.setdefault(slot.weekday, []).append(
                (slot.start_min, slot.end_min, slot.label, slot.week_parity, slot.kind)
            )

    lines = []
    for weekday in range(7):
        items = sorted(weekly.get(weekday, []), key=lambda row: (row[0], row[1]))
        value = ", ".join(
            f"{fmt_minutes(start)}–{fmt_minutes(end)}" + _slot_suffix(lang, label, parity, kind)
            for start, end, label, parity, kind in items
        ) or t(lang, "day_free")
        lines.append(f"<b>{weekday_name(lang, weekday)}</b>: {value}")
    lines.extend(sorted(dated))

    await update.effective_message.reply_text(
        t(lang, "my_schedule", schedule="\n".join(lines)), parse_mode=ParseMode.HTML
    )


async def cmd_clear(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    async with repo.session() as session:
        lang = await resolve_lang(update, session)
    keyboard = InlineKeyboardMarkup(
        [[
            InlineKeyboardButton(t(lang, "btn_yes_delete"), callback_data="clr:yes"),
            InlineKeyboardButton(t(lang, "btn_no"), callback_data="clr:no"),
        ]]
    )
    await update.effective_message.reply_text(t(lang, "clear_confirm"), reply_markup=keyboard)


async def on_clear_choice(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    await query.answer()
    async with repo.session() as session:
        lang = await resolve_lang(update, session)
        if query.data == "clr:yes":
            await repo.clear_schedule(session, update.effective_user.id)
            await session.commit()
            await query.edit_message_text(t(lang, "cleared"))
        else:
            await query.edit_message_text(t(lang, "cancelled"))


async def cmd_busy(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    async with repo.session() as session:
        await sync_user(session, update.effective_user)
        lang = await resolve_lang(update, session)
        await session.commit()
    await update.effective_message.reply_text(t(lang, "busy_ask"), parse_mode=ParseMode.HTML)
    return BUSY_WAIT


async def on_busy_text(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    message = update.effective_message
    text = message.text or ""
    async with repo.session() as session:
        lang = await resolve_lang(update, session)

    today = today_in(chat_tz(None))
    dates = _extract_dates(text, today)
    if not dates:
        await message.reply_text(t(lang, "busy_bad"), parse_mode=ParseMode.HTML)
        return BUSY_WAIT

    ranges = _find_ranges(text)
    # «12.09 весь день» и «15.09-20.09 сессия» — время не указано, значит занят целиком
    if not ranges:
        if is_all_day(text) or len(dates) > 1:
            ranges = [(0, 24 * 60, 0, 0)]
        else:
            await message.reply_text(t(lang, "busy_bad"), parse_mode=ParseMode.HTML)
            return BUSY_WAIT

    kind = find_kind(text)
    label = _busy_label(text, today)
    rendered = ", ".join(f"{fmt_minutes(s)}–{fmt_minutes(e)}" for s, e, _, _ in ranges)
    user_id = update.effective_user.id

    async with repo.session() as session:
        if len(dates) > 1:
            date_from, date_to = min(dates), max(dates)
            for start, end, _, _ in ranges:
                await repo.add_range_slot(session, user_id, date_from, date_to,
                                          start, end, label, kind)
            await session.commit()
            await message.reply_text(t(lang, "busy_range_saved",
                                       date_from=date_from.strftime("%d.%m.%Y"),
                                       date_to=date_to.strftime("%d.%m.%Y"), time=rendered))
        else:
            for start, end, _, _ in ranges:
                await repo.add_dated_slot(session, user_id, dates[0], start, end, label, kind)
            await session.commit()
            await message.reply_text(t(lang, "busy_saved",
                                       date=dates[0].strftime("%d.%m.%Y"), time=rendered))
    return ConversationHandler.END


def _extract_dates(text: str, today) -> list:
    """Все даты в сообщении. Две даты = диапазон (сессия, поездка)."""
    found = []
    for token in re.split(r"[\s,;]+|(?<=\d)[–—](?=\d)|(?<=\d)-(?=\d{1,2}[./-])", text):
        day = parse_date_token(token, today)
        if day and day not in found:
            found.append(day)
    return found


def _busy_label(text: str, today) -> str:
    """Название занятости: всё, что не дата и не диапазон времени."""
    words: list[str] = []
    for token in re.split(r"[\s,;]+", text):
        if not token:
            continue
        if parse_date_token(token, today):
            continue
        if re.fullmatch(r"[\d:.\-–—/hч]+", token):   # даты-диапазоны и времена
            continue
        words.append(token)
    return " ".join(words).strip(" .:,;-–—")[:60]


async def cancel(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    async with repo.session() as session:
        lang = await resolve_lang(update, session)
    context.user_data.pop("pending_schedule", None)
    context.user_data.pop("wiz_day", None)
    await update.effective_message.reply_text(t(lang, "cancelled"))
    return ConversationHandler.END


def register(app) -> None:
    import_conv = ConversationHandler(
        entry_points=[
            CommandHandler("schedule", cmd_schedule, filters=PRIVATE),
            CommandHandler("import", cmd_schedule, filters=PRIVATE),
            MessageHandler(PRIVATE & filters.TEXT & ~filters.COMMAND, on_schedule_text),
        ],
        states={
            IMPORT_WAIT: [MessageHandler(PRIVATE & filters.TEXT & ~filters.COMMAND, on_schedule_text)],
            IMPORT_CONFIRM: [CallbackQueryHandler(on_schedule_confirm, pattern=r"^sch:")],
        },
        fallbacks=[CommandHandler("cancel", cancel)],
        name="schedule_import",
        persistent=True,
        conversation_timeout=15 * 60,
    )

    wizard_conv = ConversationHandler(
        entry_points=[CommandHandler("wizard", cmd_wizard, filters=PRIVATE)],
        states={
            WIZ_PICK_DAY: [CallbackQueryHandler(on_wizard_day, pattern=r"^wiz:")],
            WIZ_WAIT_TEXT: [MessageHandler(PRIVATE & filters.TEXT & ~filters.COMMAND, on_wizard_text)],
        },
        fallbacks=[CommandHandler("cancel", cancel)],
        name="schedule_wizard",
        persistent=True,
        conversation_timeout=15 * 60,
    )

    busy_conv = ConversationHandler(
        entry_points=[CommandHandler("busy", cmd_busy, filters=PRIVATE)],
        states={BUSY_WAIT: [MessageHandler(PRIVATE & filters.TEXT & ~filters.COMMAND, on_busy_text)]},
        fallbacks=[CommandHandler("cancel", cancel)],
        name="schedule_busy",
        persistent=True,
        conversation_timeout=10 * 60,
    )

    # Мастер и разовая занятость идут раньше свободного импорта:
    # иначе их текстовые шаги перехватит entry_point импорта.
    app.add_handler(wizard_conv)
    app.add_handler(busy_conv)
    app.add_handler(import_conv)
    app.add_handler(CommandHandler("myschedule", cmd_my_schedule, filters=PRIVATE))
    app.add_handler(CommandHandler("clear", cmd_clear, filters=PRIVATE))
    app.add_handler(CallbackQueryHandler(on_clear_choice, pattern=r"^clr:"))

File: src/bot/handlers/schedule.ts
/** Ввод расписания в личке: /schedule, /import, /wizard, /myschedule, /clear, /busy. */

import "server-only";

import { fmtMinutes } from "@/core/intervals";
import {
  type ParseResult,
  findKind,
  findRangesWithPos,
  isAllDay,
  parseAny,
  parseResultOk,
} from "@/core/parser";
import { escapeHtml } from "@/core/textutils";
import {
  type DateStr,
  compareDates,
  formatDM,
  formatDMY,
  parseDateToken,
  todayIn,
} from "@/core/timeutils";
import * as repo from "@/db/repo";
import { DEFAULT_TZ } from "@/core/timeutils";
import { t, weekdayName } from "@/i18n";
import {
  answerCallbackQuery,
  editMessageText,
  keyboard,
  sendMessage,
  type TgCallbackQuery,
  type TgMessage,
} from "../api";
import {
  type SlotTuple,
  clearPrivateState,
  getPrivateState,
  resolveLang,
  setPrivateState,
  syncUser,
} from "../context";

// --------------------------------------------------------------------------
// Импорт расписания текстом
// --------------------------------------------------------------------------

export async function cmdSchedule(message: TgMessage): Promise<void> {
  if (message.from) {
    await syncUser(message.from);
    await clearPrivateState(message.from.id);
  }
  const lang = await resolveLang(message.chat, message.from);
  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "schedule_intro"),
    parse_mode: "HTML",
  });
}

function slotSuffix(lang: string, label: string, parity: number | null, kind: string): string {
  const parts: string[] = [];
  if (label) parts.push(escapeHtml(label));
  if (parity !== null) parts.push(t(lang, parity === 0 ? "parity_odd" : "parity_even"));
  if (kind && kind !== "class") parts.push(t(lang, `kind_${kind}`));
  return parts.length ? ` <i>(${parts.join(", ")})</i>` : "";
}

function renderParsePreview(result: ParseResult, lang: string): string {
  const byDay = new Map<number, SlotTuple[]>();
  for (const slot of result.slots) {
    const list = byDay.get(slot.weekday) ?? [];
    list.push({
      weekday: slot.weekday,
      start: slot.startMin,
      end: slot.endMin,
      label: slot.label,
      parity: slot.parity,
      kind: slot.kind,
    });
    byDay.set(slot.weekday, list);
  }

  const lines: string[] = [];
  for (let weekday = 0; weekday < 7; weekday += 1) {
    const items = byDay.get(weekday);
    if (items) {
      const sorted = [...items].sort((a, b) => a.start - b.start || a.end - b.end);
      const rendered = sorted
        .map(
          (item) =>
            `${fmtMinutes(item.start)}–${fmtMinutes(item.end)}` +
            slotSuffix(lang, item.label, item.parity, item.kind),
        )
        .join(", ");
      lines.push(`<b>${weekdayName(lang, weekday)}</b>: ${rendered}`);
    } else if (result.freeDays.includes(weekday)) {
      lines.push(`<b>${weekdayName(lang, weekday)}</b>: ${t(lang, "day_free")}`);
    }
  }
  return lines.join("\n") || "—";
}

/** Свободный текст в личке — это и есть импорт расписания. */
export async function onScheduleText(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;

  await syncUser(from);
  const lang = await resolveLang(message.chat, from);
  const result = parseAny(message.text ?? "");

  if (!parseResultOk(result)) {
    await sendMessage({
      chat_id: message.chat.id,
      text: t(lang, "parse_failed"),
      parse_mode: "HTML",
    });
    return;
  }

  await setPrivateState(from.id, {
    mode: "import_confirm",
    slots: result.slots.map((slot) => ({
      weekday: slot.weekday,
      start: slot.startMin,
      end: slot.endMin,
      label: slot.label,
      parity: slot.parity,
      kind: slot.kind,
    })),
    freeDays: result.freeDays,
  });

  let body = t(lang, "parse_preview", { schedule: renderParsePreview(result, lang) });
  if (result.errors.length) {
    body += t(lang, "parse_errors", {
      lines: escapeHtml(result.errors.slice(0, 5).join("\n")),
    });
  }

  await sendMessage({
    chat_id: message.chat.id,
    text: body,
    parse_mode: "HTML",
    reply_markup: keyboard([
      [
        { text: t(lang, "btn_confirm"), callback_data: "sch:ok" },
        { text: t(lang, "btn_retry"), callback_data: "sch:retry" },
      ],
    ]),
  });
}

export async function onScheduleConfirm(query: TgCallbackQuery): Promise<void> {
  const message = query.message;
  if (!message) return;
  await answerCallbackQuery({ callback_query_id: query.id });

  const action = (query.data ?? "").split(":")[1];
  const lang = await resolveLang(message.chat, query.from);

  if (action === "retry") {
    await clearPrivateState(query.from.id);
    await editMessageText({
      chat_id: message.chat.id,
      message_id: message.message_id,
      text: t(lang, "schedule_intro"),
      parse_mode: "HTML",
    });
    return;
  }

  const state = await getPrivateState(query.from.id);
  if (state?.mode !== "import_confirm") {
    await editMessageText({
      chat_id: message.chat.id,
      message_id: message.message_id,
      text: t(lang, "error_generic"),
    });
    return;
  }
  await clearPrivateState(query.from.id);

  const touched = [...new Set([...state.slots.map((s) => s.weekday), ...state.freeDays])].sort(
    (a, b) => a - b,
  );
  await repo.replaceWeeklySlots(query.from.id, touched, state.slots, "import");

  await editMessageText({
    chat_id: message.chat.id,
    message_id: message.message_id,
    text: t(lang, "schedule_saved"),
  });
}

// --------------------------------------------------------------------------
// Мастер по дням
// --------------------------------------------------------------------------

function dayKeyboard(lang: string) {
  const rows: { text: string; callback_data: string }[][] = [];
  let row: { text: string; callback_data: string }[] = [];
  for (let weekday = 0; weekday < 7; weekday += 1) {
    row.push({ text: weekdayName(lang, weekday).slice(0, 3), callback_data: `wiz:day:${weekday}` });
    if (row.length === 4) {
      rows.push(row);
      row = [];
    }
  }
  if (row.length) rows.push(row);
  rows.push([{ text: t(lang, "btn_done"), callback_data: "wiz:done" }]);
  return keyboard(rows);
}

export async function cmdWizard(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;
  await syncUser(from);
  const lang = await resolveLang(message.chat, from);
  await setPrivateState(from.id, { mode: "wizard", weekday: null });
  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "wizard_pick_day"),
    reply_markup: dayKeyboard(lang),
  });
}

export async function onWizardDay(query: TgCallbackQuery): Promise<void> {
  const message = query.message;
  if (!message) return;
  await answerCallbackQuery({ callback_query_id: query.id });
  const lang = await resolveLang(message.chat, query.from);

  if (query.data === "wiz:done") {
    await repo.markFilled(query.from.id, true);
    await clearPrivateState(query.from.id);
    await editMessageText({
      chat_id: message.chat.id,
      message_id: message.message_id,
      text: t(lang, "wizard_done"),
    });
    return;
  }

  const weekday = Number((query.data ?? "").split(":").pop());
  await setPrivateState(query.from.id, { mode: "wizard", weekday });
  await editMessageText({
    chat_id: message.chat.id,
    message_id: message.message_id,
    text: t(lang, "wizard_ask_day", { day: weekdayName(lang, weekday) }),
    parse_mode: "HTML",
  });
}

export async function onWizardText(message: TgMessage, weekday: number | null): Promise<void> {
  const from = message.from;
  if (!from) return;
  const lang = await resolveLang(message.chat, from);

  if (weekday === null) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "error_generic") });
    await clearPrivateState(from.id);
    return;
  }

  const ranges = findRangesWithPos(message.text ?? "");
  const slots = ranges.map((range) => ({
    weekday,
    start: range.start,
    end: range.end,
    label: "",
    parity: null,
    kind: "class",
  }));
  await repo.replaceWeeklySlots(from.id, [weekday], slots, "wizard");
  await setPrivateState(from.id, { mode: "wizard", weekday: null });

  const rendered =
    slots.map((slot) => `${fmtMinutes(slot.start)}–${fmtMinutes(slot.end)}`).join(", ") ||
    t(lang, "day_free");

  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "wizard_day_saved", { day: weekdayName(lang, weekday), slots: rendered }),
  });
  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "wizard_pick_day"),
    reply_markup: dayKeyboard(lang),
  });
}

// --------------------------------------------------------------------------
// Просмотр, очистка, разовая занятость
// --------------------------------------------------------------------------

export async function cmdMySchedule(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;
  await syncUser(from);
  const lang = await resolveLang(message.chat, from);
  const slots = await repo.getSlots(from.id);

  if (slots.length === 0) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "schedule_empty") });
    return;
  }

  const weekly = new Map<number, SlotTuple[]>();
  const dated: string[] = [];
  for (const slot of slots) {
    if (slot.specificDate !== null) {
      dated.push(
        `📌 <b>${formatDMY(slot.specificDate)}</b>: ` +
          `${fmtMinutes(slot.startMin)}–${fmtMinutes(slot.endMin)}` +
          slotSuffix(lang, slot.label, null, slot.kind),
      );
    } else if (slot.dateFrom !== null && slot.dateTo !== null) {
      dated.push(
        `📌 <b>${formatDM(slot.dateFrom)}–${formatDMY(slot.dateTo)}</b>: ` +
          `${fmtMinutes(slot.startMin)}–${fmtMinutes(slot.endMin)}` +
          slotSuffix(lang, slot.label, null, slot.kind),
      );
    } else if (slot.weekday !== null) {
      const list = weekly.get(slot.weekday) ?? [];
      list.push({
        weekday: slot.weekday,
        start: slot.startMin,
        end: slot.endMin,
        label: slot.label,
        parity: slot.weekParity,
        kind: slot.kind,
      });
      weekly.set(slot.weekday, list);
    }
  }

  const lines: string[] = [];
  for (let weekday = 0; weekday < 7; weekday += 1) {
    const items = [...(weekly.get(weekday) ?? [])].sort(
      (a, b) => a.start - b.start || a.end - b.end,
    );
    const value =
      items
        .map(
          (item) =>
            `${fmtMinutes(item.start)}–${fmtMinutes(item.end)}` +
            slotSuffix(lang, item.label, item.parity, item.kind),
        )
        .join(", ") || t(lang, "day_free");
    lines.push(`<b>${weekdayName(lang, weekday)}</b>: ${value}`);
  }
  lines.push(...[...dated].sort());

  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "my_schedule", { schedule: lines.join("\n") }),
    parse_mode: "HTML",
  });
}

export async function cmdClear(message: TgMessage): Promise<void> {
  const lang = await resolveLang(message.chat, message.from);
  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "clear_confirm"),
    reply_markup: keyboard([
      [
        { text: t(lang, "btn_yes_delete"), callback_data: "clr:yes" },
        { text: t(lang, "btn_no"), callback_data: "clr:no" },
      ],
    ]),
  });
}

export async function onClearChoice(query: TgCallbackQuery): Promise<void> {
  const message = query.message;
  if (!message) return;
  await answerCallbackQuery({ callback_query_id: query.id });
  const lang = await resolveLang(message.chat, query.from);

  if (query.data === "clr:yes") {
    await repo.clearSchedule(query.from.id);
    await editMessageText({
      chat_id: message.chat.id,
      message_id: message.message_id,
      text: t(lang, "cleared"),
    });
  } else {
    await editMessageText({
      chat_id: message.chat.id,
      message_id: message.message_id,
      text: t(lang, "cancelled"),
    });
  }
}

export async function cmdBusy(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;
  await syncUser(from);
  const lang = await resolveLang(message.chat, from);
  await setPrivateState(from.id, { mode: "busy" });
  await sendMessage({ chat_id: message.chat.id, text: t(lang, "busy_ask"), parse_mode: "HTML" });
}

// Разделяем и по «12.09-15.09»: дефис между датами — не часть токена.
const BUSY_SEPARATOR = /[\s,;]+|(?<=\d)[–—](?=\d)|(?<=\d)-(?=\d{1,2}[./-])/gu;

type Token = { value: string; start: number; end: number };

function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  const separator = new RegExp(BUSY_SEPARATOR.source, BUSY_SEPARATOR.flags);
  let cursor = 0;
  let match: RegExpExecArray | null;
  while ((match = separator.exec(text)) !== null) {
    tokens.push({ value: text.slice(cursor, match.index), start: cursor, end: match.index });
    cursor = match.index + match[0].length;
  }
  tokens.push({ value: text.slice(cursor), start: cursor, end: text.length });
  return tokens.filter((token) => token.value !== "");
}

/**
 * Даты сообщения и текст без них. Две даты = диапазон (сессия, поездка).
 *
 * Даты вырезаются из текста, потому что «15.09-20.09» само по себе выглядит
 * как диапазон времени 15:09–20:09: точка в разборе времени равноправна
 * двоеточию. Без этой маски «15.09-20.09 сессия» записывалось бы не сессией
 * на неделю, а часом с небольшим каждый день.
 */
function extractDates(
  text: string,
  today: DateStr,
): { dates: DateStr[]; withoutDates: string } {
  const dates: DateStr[] = [];
  let masked = text;

  for (const token of tokenize(text)) {
    const day = parseDateToken(token.value, today);
    if (!day) continue;
    if (!dates.includes(day)) dates.push(day);
    // Замена той же длины — позиции остальных токенов не съезжают.
    masked =
      masked.slice(0, token.start) +
      " ".repeat(token.end - token.start) +
      masked.slice(token.end);
  }
  return { dates, withoutDates: masked };
}

/** Название занятости: всё, что не дата и не диапазон времени. */
function busyLabel(text: string, today: DateStr): string {
  const words: string[] = [];
  for (const token of text.split(/[\s,;]+/u)) {
    if (!token) continue;
    if (parseDateToken(token, today)) continue;
    if (/^[\d:.\-–—/hч]+$/u.test(token)) continue; // даты-диапазоны и времена
    words.push(token);
  }
  let label = words.join(" ");
  const trim = " .:,;-–—";
  let start = 0;
  let end = label.length;
  while (start < end && trim.includes(label[start])) start += 1;
  while (end > start && trim.includes(label[end - 1])) end -= 1;
  label = label.slice(start, end);
  return label.slice(0, 60);
}

export async function onBusyText(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;
  const lang = await resolveLang(message.chat, from);
  const text = message.text ?? "";
  const today = todayIn(DEFAULT_TZ);

  const { dates, withoutDates } = extractDates(text, today);
  if (dates.length === 0) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "busy_bad"), parse_mode: "HTML" });
    return;
  }

  let ranges = findRangesWithPos(withoutDates).map((range) => ({
    start: range.start,
    end: range.end,
  }));
  // «12.09 весь день» и «15.09-20.09 сессия» — время не указано, значит занят целиком.
  if (ranges.length === 0) {
    if (isAllDay(text) || dates.length > 1) {
      ranges = [{ start: 0, end: 24 * 60 }];
    } else {
      await sendMessage({
        chat_id: message.chat.id,
        text: t(lang, "busy_bad"),
        parse_mode: "HTML",
      });
      return;
    }
  }

  const kind = findKind(text);
  const label = busyLabel(text, today);
  const rendered = ranges
    .map((range) => `${fmtMinutes(range.start)}–${fmtMinutes(range.end)}`)
    .join(", ");

  await clearPrivateState(from.id);

  if (dates.length > 1) {
    const sorted = [...dates].sort(compareDates);
    const dateFrom = sorted[0];
    const dateTo = sorted[sorted.length - 1];
    for (const range of ranges) {
      await repo.addRangeSlot({
        userId: from.id,
        dateFrom,
        dateTo,
        start: range.start,
        end: range.end,
        label,
        kind,
      });
    }
    await sendMessage({
      chat_id: message.chat.id,
      text: t(lang, "busy_range_saved", {
        date_from: formatDMY(dateFrom),
        date_to: formatDMY(dateTo),
        time: rendered,
      }),
    });
    return;
  }

  for (const range of ranges) {
    await repo.addDatedSlot({
      userId: from.id,
      day: dates[0],
      start: range.start,
      end: range.end,
      label,
      kind,
    });
  }
  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "busy_saved", { date: formatDMY(dates[0]), time: rendered }),
  });
}
